cgroup_poller.go

```diff
@@ -0,0 +1,99 @@
+package shh
+
+import (
+	"io/ioutil"
+	"strings"
+	"time"
+
+	"github.com/heroku/shh/Godeps/_workspace/src/github.com/freeformz/filechan"
+)
+
+const (
+	CGROUPS_PATH = "/sys/fs/cgroup"
+)
+
+type Cgroup struct {
+	measurements chan<- Measurement
+	last         map[string]uint64
+	cgroups      []string
+	// The kernel will report CPU usage in centiseconds.  This
+	// stores the total centiseconds in the polling interval.
+	totalCentis uint64
+}
+
+func NewCgroupPoller(measurements chan<- Measurement, config Config) Cgroup {
+	return Cgroup{
+		measurements: measurements,
+		last:         make(map[string]uint64),
+		cgroups:      config.Cgroups,
+		// convert the interval to centiseconds
+		totalCentis: uint64(config.Interval.Nanoseconds() / 10000000),
+	}
+}
+
+func sanitizeMetricName(name string) string {
+	return strings.Replace(strings.Replace(name, ".", "-", -1), "/", "--", -1)
+}
+
+// parsePercent parses a line from cpuacct.stat like "user 12345678",
+// calculates the delta from the last measurement, calculates the
+// average percentage of one CPU core used, and submits the data point.
+func (poller Cgroup) parsePercentCpu(line string, tick time.Time, cgroup string) {
+	fields := strings.Fields(line)
+
+	// "user" or "system"
+	metric := fields[0]
+
+	// absolute number of centiseconds of CPU time used
+	centis := Atouint64(fields[1])
+
+        key := cgroup + "." + metric
+	last, exists := poller.last[key]
+
+	if exists {
+		delta := centis - last
+		percent := float64(delta) * 100.0 / float64(poller.totalCentis)
+
+		poller.measurements <- FloatGaugeMeasurement{tick, poller.Name(), []string{sanitizeMetricName(cgroup), "cpu", metric}, percent, Percent}
+	}
+
+	poller.last[key] = centis
+}
+
+func (poller Cgroup) parseMaxMemory(metric string, fileName string, tick time.Time, cgroup string) {
```

Comment:
This function "emits" the metric as well. Maybe have this return the created FloatGaugeMeasurement, or even just change the name to `emitMaxMemory` so that it speaks to what it does?